fix(upgrade): call ng1 lifecycle hooks

bradyisom opened the pull request:
**Please check if the PR fulfills these requirements**
- [ x ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ x ] Tests for the changes have been added (for bug fixes / features)
- [ x ] Docs have been added / updated (for bug fixes / features)


**What kind of change does this PR introduce?** (check one with "x")
```
[ x ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

**What is the current behavior?** (You can also link to an open issue here)
Angular 1.5 components with lifecycle hooks consumed by Angular 2 components via upgradeNg1Component never get `$doCheck`, `$onChanges` or `$onDestroy` called. 


**What is the new behavior?**
Angular 1.5 components get `$doCheck`, `$onChanges` or `$onDestroy` called at the appropriate points, if they the component defines them.


**Does this PR introduce a breaking change?** (check one with "x")
```
[ ] Yes
[ x ] No
```

If this PR contains a breaking change, please describe the impact and migration path for existing applications: ...


**Other information**:

bradyisom commented:
Any comments or direction on this? This is a fairly small change, in my opinion. The code without this change gives me no hooks to have my ng1 lifecycle hooks called. The really bad one is `$onDestroy`. Without it, my ng1 components have no way to clean up their resources.

petebacondarwin commented on modules/@angular/upgrade/src/upgrade_ng1_adapter.ts:
```diff
@@ -262,12 +263,18 @@ class UpgradeNg1ComponentAdapter implements OnInit, OnChanges, DoCheck {
         const change: SimpleChange = changes[name];
         this.setComponentProperty(name, change.currentValue);
+        ng1Changes = ng1Changes || {};
+        ng1Changes[this.propertyMap[name]] = change;
```
It's a good job that we copied the API of the `SimpleChanges` class into the objects used for `$onChanges` :-)

petebacondarwin commented on modules/@angular/upgrade/src/upgrade_ng1_adapter.ts:
```diff
@@ -262,12 +263,18 @@ class UpgradeNg1ComponentAdapter implements OnInit, OnChanges, DoCheck {
+        ng1Changes[this.propertyMap[name]] = change;
       }
     }
+    if (ng1Changes && this.destinationObj.$onChanges) {
```
I guess that `ngOnChanges` is never called with an empty `changes` parameter. So perhaps we shouldn't need to do this check?

bradyisom replied:
Does that mean we are guaranteed to get into the `hasOwnProperty` if statement at least once? If so, then I agree that we could remove this check.

petebacondarwin replied:
I believe that this is the case.

petebacondarwin commented on modules/@angular/upgrade/src/upgrade_ng1_adapter.ts:
```diff
@@ -262,12 +263,18 @@ class UpgradeNg1ComponentAdapter implements OnInit, OnChanges, DoCheck {
 
   ngOnChanges(changes: SimpleChanges) {
+    let ng1Changes: any = null;
     for (const name in changes) {
```
If we used `Object.keys` here I don't think we would need to do the `hasOwnProperty` check (right?), which would simplify this whole method.

bradyisom replied:
The `hasOwnProperty` code was already there to detect what to call `setComponentProperty` on.

petebacondarwin replied:
Yes, I understand that the code was there already and so it is not your fault @bradyisom .
I don't mind if you don't change it in this PR. If you do then I would do it via a separate `refactor` commit.

petebacondarwin commented on modules/@angular/upgrade/src/upgrade_ng1_adapter.ts:
```diff
@@ -262,12 +263,18 @@ class UpgradeNg1ComponentAdapter implements OnInit, OnChanges, DoCheck {
+    }
   }
 
   ngDoCheck(): number {
```
`ngDoCheck`'s  return type is actually `void`. So there is no need to do this counting. In fact the `count` variable is even set as `const`(!) and so this is always returning `0` :-)

bradyisom replied:
I'm assuming you are commenting on the return value from `ngDoCheck`. That code already existed before this PR. Are you saying you want me to clean up the return value of that function as part of this same PR? I can...

petebacondarwin commented on modules/@angular/upgrade/test/upgrade_spec.ts:
```diff
@@ -922,6 +922,106 @@ export function main() {
+             expect(multiTrim(document.body.textContent)).toEqual(valueToFind);
+             ref.dispose();
+           });
+         }));
```
I wonder if we should also test that `doCheck` is called subsequently after interactions as well as straight after initialization?

petebacondarwin replied:
And similarly for `onChanges`?

petebacondarwin commented on modules/@angular/upgrade/test/upgrade_spec.ts:
```diff
@@ -922,6 +922,106 @@ export function main() {
+             bindings: {},
+             template: '<div>ng1</div>',
+             controller: function($rootScope: any) {
+               this.$onDestroy = function() { $rootScope.$broadcast('destroyed'); };
```
Couldn't we use a spy here? It would seem to be easier than having to listen to events on the scope.

petebacondarwin commented on modules/@angular/upgrade/test/upgrade_spec.ts:
```diff
@@ -922,6 +922,106 @@ export function main() {
+           const element = html(`<div ng-if="!destroy"><ng2></ng2></div>`);
+           adapter.bootstrap(element, ['ng1']).ready((ref) => {
+             (<any>ref.ng1RootScope).destroy = false;
+             setTimeout(() => { (<any>ref.ng1RootScope).destroy = true; }, 100);
```
And perhaps we could use `fakeAsync` to control this test more accurately?
But perhaps I am being too picky?

petebacondarwin reviewed (approved):
A few minor suggestions but generally LGTM

vicb commented:
@petebacondarwin what is the status here ?
add the "merge" label if is it ready to merge or the "cleanup" label otherwise. Thanks.

petebacondarwin commented:
The only changes that I would like to see before I would be happy to merge this are the additional tests that check that hooks are called after interactions.
The other comments are optional.

bradyisom committed:
fix(upgrade): call ng1 lifecycle hooks

bradyisom commented:
I added tests for hooks called after interactions. 
I refactored some tests to use spies.
I did not change to use fakeAsync. I continued using `setTimeout` like other tests in the same file.
I added a refactor commit to clean up the implementation of `ngOnChanges` and `ngDoCheck`.

bradyisom commented:
All checks have passed.

petebacondarwin commented:
LGTM - thanks @bradyisom - let's get this merged.

gkalpak commented on modules/@angular/upgrade/src/upgrade_ng1_adapter.ts:
```diff
@@ -287,7 +290,15 @@ class UpgradeNg1ComponentAdapter implements OnInit, OnChanges, DoCheck {
       }
     }
-    return count;
+    if (this.destinationObj.$doCheck) {
```
`$doCheck` should only run on controllers only, not scopes.

gkalpak replied:
Not related to this PR, but the `$onInit()` hook has the same issue. (It has even more issues, because it is called after the linking phase, which is not what happens in ng1.)

gkalpak commented on modules/@angular/upgrade/src/upgrade_ng1_adapter.ts:
```diff
@@ -287,7 +290,15 @@ class UpgradeNg1ComponentAdapter implements OnInit, OnChanges, DoCheck {
+  }
+
+  ngOnDestroy() {
+    if (this.destinationObj.$onDestroy) {
```
`$onDestroy` should run on controller only, not scopes.

gkalpak commented:
Since yu are at it, you might want to add support for the `$postLink` hook as well (but not necessary).

gkalpak commented:
Oops, I just realized this has the `pr_action: merge` label. Too late. We can fix the hooks in subsequent PRs :smiley:

petebacondarwin commented:
I removed the merge label - can you not do that @gkalpak?
@bradyisom can you address comments about only calling on controllers and not on scope?
Let's fix the comment about when $onInit gets called in another PR.

chuckjaz commented:
You have time to fix this since I have not merged this yet. I will wait until the merge label is back.

bradyisom commented:
I can fix to only call on controllers for the `$onDestroy`, `$onChanges` and `$doCheck` calls. I could also add the same check for the `$onInit` call. However, if there are other issues, I'll leave that one alone.

gkalpak commented:
Except that `$onChanges` is a little special: It gets called on the binding destination (controller _or_ scope), not necessarily on the controller. The other hooks get called on controllers only (and regardless of whether they are the binding destination or not).

For reference, here is how `upgrade/static` handles hooks: [`UpgradeComponent][1]
(It is currently missing `$doCheck()` though.)

--
Feel free to take this one as far as you like and I will be happy to take it from there. It is still much better than what we have now and there are issues in the the current implementation anyway.

[1]: https://github.com/angular/angular/blob/f340e1a4144ff09882daaf6266b34bd866fdb0b7/modules/%40angular/upgrade/src/aot/upgrade_component.ts

bradyisom commented:
@gkalpak, so you are saying that I should remove the `this.directive.controller` check from the `$onChanges` hook? I can do that and then leave the other refactoring to you, or somebody else.

bradyisom commented:
All checks are passing again. I think this is ready for merge.

petebacondarwin commented on modules/@angular/upgrade/src/upgrade_ng1_adapter.ts:
```diff
@@ -262,16 +263,18 @@ class UpgradeNg1ComponentAdapter implements OnInit, OnChanges, DoCheck {
-        const change: SimpleChange = changes[name];
-        this.setComponentProperty(name, change.currentValue);
-      }
+    let ng1Changes: any = {};
```
should be `const`

bradyisom commented:
Changed let -> const

bradyisom commented:
Are we ready to merge this one? I think I'm done with the suggestions.

petebacondarwin commented:
Thanks @bradyisom . Let's merge this guy

bradyisom commented:
Thank you, gentlemen.

chuckjaz commented:
Please squash the commit so we are not commit a clean-up commit with the commit it cleans up.

bradyisom commented:
squash complete
